src/daemon/rpmostreed-os.cxx

```diff
@@ -1138,6 +1140,92 @@ os_handle_get_packages (RPMOSTreeOS *interface, GDBusMethodInvocation *invocatio
   return TRUE;
 }
 
+static void
+query_results_to_builder (HyQuery query, GVariantBuilder *builder, const gchar *id)
+{
+  g_autoptr (GPtrArray) pkglist = NULL;
+  pkglist = hy_query_run (query);
+  for (guint i = 0; i < pkglist->len; i++)
+    {
+      auto pkg = static_cast<DnfPackage *> (g_ptr_array_index (pkglist, i));
+      os_add_package_info_to_builder (pkg, builder, id);
+    }
+}
+
+static void
+search_packages_by_filter (HyQuery query, GVariantBuilder *builder, const gchar *const *names,
+                           int *keynames, const gchar *id)
+{
+  hy_query_clear (query);
+  for (guint j = 0; keynames[j] != -1; j++)
+    {
+      for (guint i = 0; names[i] != NULL; i++)
+        {
+          if (!hy_is_glob_pattern (names[i]))
+            {
+              hy_query_filter (query, keynames[j], HY_EQ | HY_ICASE, names[i]);
+            }
+          else
+            {
+              hy_query_filter (query, keynames[j], HY_GLOB | HY_ICASE, names[i]);
+            }
+        }
+    }
+  query_results_to_builder (query, builder, id);
+
+  hy_query_clear (query);
+  for (guint j = 0; keynames[j] != -1; j++)
+    {
+      for (guint i = 0; names[i] != NULL; i++)
+        {
+          if (!hy_is_glob_pattern (names[i]))
+            {
+              hy_query_filter (query, keynames[j], HY_SUBSTR | HY_ICASE, names[i]);
+            }
+          else
+            {
+              hy_query_filter (query, keynames[j], HY_GLOB | HY_ICASE, names[i]);
+            }
+        }
+    }
+  query_results_to_builder (query, builder, id);
+}
+
+static gboolean
+os_handle_search (RPMOSTreeOS *interface, GDBusMethodInvocation *invocation,
+                  const gchar *const *names)
+{
+  GError *local_error = NULL;
+  g_autoptr (GCancellable) cancellable = NULL;
+  g_autoptr (DnfContext) dnfctx = NULL;
+
+  sd_journal_print (LOG_INFO, "Handling Search for caller %s",
+                    g_dbus_method_invocation_get_sender (invocation));
+
+  dnfctx = os_create_dnf_context_simple (interface, TRUE, cancellable, &local_error);
+  if (dnfctx == NULL)
+    return os_throw_dbus_invocation_error (invocation, &local_error);
+
+  hy_autoquery HyQuery query = hy_query_create (dnf_context_get_sack (dnfctx));
+
+  GVariantBuilder builder;
+  g_variant_builder_init (&builder, (const GVariantType *)"aa{sv}");
+
+  int keynames_a[3] = { HY_PKG_NAME, HY_PKG_SUMMARY, -1 };
+  search_packages_by_filter (query, &builder, names, keynames_a, "Name & Summary");
```

Comment:
I would omit the human-readable text "Name & Summary" from the DBus API, and instead print it on the client.  We'd just probably document these as "match groups" maybe?

One thing to bear in mind here is that there are other clients for the DBus API - notably [Cockpit](https://cockpit-project.org/).  Related to human-readable text I think Cockpit is translated, whereas rpm-ostree is not (today).

(See e.g. https://bugzilla.redhat.com/show_bug.cgi?id=2209924 for a bug they filed recently, not related to this directly; just showing as an example of an external project using the DBus API)